README
PcapWriter
Perl-Library um mal schnell ein Pcap file mit ein paar TCP-Verbindungen
drin zu erstellen:

 | my $w = Net::PcapWriter->new('test.pcap');
 | my $conn = $w->tcp_conn('1.2.3.4',1234,'5.6.7.8',80) or die;
 | $conn->write(0,"GET / HTTP/1.0 200 Ok\r\n\r\n");
 | $conn->write(1,"HTTP/1.0 12345 ab\r\nContent-length: 0\r\n\r\n");



Changes
0.723 2016-01-29
- fixed off-by-one calculation of sequence number for FIN in TCP.pm
- added explicit connect function in TCP.pm
0.722 2016-01-21
- reopen function usable for rotating output etc
0.721 2016-01-08
- more control over the closing process with the close method in TCP
0.720 2015-12-16
- option to disable checksum calculation completly which gives a major
  performance boost
0.719 2015-12-15
- faster checksum calculation which has a major performance impact
0.718 2015-12-08
- fix IPv6 support (was bad ethernet type id)
- performance should be faster for TCP and UDP now
- Behavior change: Net::PcapWriter::packet gets Layer 2 packet now
  (ethernet layer) instead of IP packet
0.717 2015-12-07
- update documentation for TCP ack, shutdown to include timestamp.
  Add write_with_flags to documented API for TCP. Make write_with_flags
  update the internal state of the connection. Optimize DESTROY/_close
  to write only the packets which are necessary to close the connection.
0.716 2015-08-31
- set binmode for output file to work on Windows platforms.
  Thanks to security[AT]knochner[DOT]com for reporting.
0.715 2015-02-02
- merge changes to fix kwalitee issues.
  Thanks to oeuftete and to CPAN Pull Request Challenge
0.714 2014-12-15
- forgot to add ICMP_Echo.pm to MANIFEST
0.713 2014-12-14
- support for ICMP/ICMPv6 echo, based on patch from EmmanuelBau
0.712 2014-07-04
- fix checksum computation, thanks to BLUHM for reporting
0.711 2014-07-03
- added support for IPv6
0.710_2 2013-07-04
- try to find usable inet_pton in IP.pm and export it for TCP and UDP
0.710_1 2013-07-02
- try to detect unusable inet_pton in Makefile.PL (Win32)
0.71 2013-05-07
- fixed writing timestamp in Tcp.pm and Udp.pm
  Thanks to christian[DOT]janssen[AT]evonik[DOT]com for reporting
0.7 2013-03-06
- add udp_conn
0.6 2012-10-01
- add license to Makefile.PL, change email in author section
0.5 2012-03-01
- forgot to add PREREQ_PM to WriteMakefile
0.4 2012-02-29
- add Socket6 to PREREQ_PM if inet_pton is not in Socket yet (<5.12)
0.3 2012-02-27
- switch to DLT_EN10MB instead of DLT_RAW, because type-id is the same for
  all systems
0.2 2011-10-20
- make it work on 5.10 and 5.8, where inet_pton is in Socket6
  and not Socket
0.1 2011-10-07
- initial release


lib/Net/PcapWriter/IP.pm
use strict;
use warnings;
package Net::PcapWriter::IP;
use Socket qw(AF_INET AF_INET6);

use base 'Exporter';
# re-export the usable inet_pton
our @EXPORT = qw(ip_chksum ip4_packet ip6_packet ip_packet inet_pton);

BEGIN { 
    # inet_pton is in Socket since 5.12
    # but even if it is in Socket it can throw a non-implemented error
    eval {
	Socket->import('inet_pton');
	inet_pton(AF_INET,'127.0.0.1');
	inet_pton(AF_INET6,'::1');
	1
    } or eval {
	require Socket6;
	Socket6->import('inet_pton');
	inet_pton(AF_INET,'127.0.0.1');
	inet_pton(AF_INET6,'::1');
	1
    } or die "you need either a modern perl or Socket6"
}


# construct IPv4 packet or packet generating sub
sub ip4_packet {
    my ($data,$src,$dst,$protocol,$chksum_offset,$no_pseudo_header) = @_;
    my $hdr = pack('CCnnnCCna4a4',
	0x45,             # version 4, len=5 (no options)
	0,                # type of service
	defined($data) ? length($data)+20 : 20, # total length
	0,0,              # id=0, not fragmented
	128,              # TTL
	$protocol,
	0,                # checksum - computed later
	scalar(inet_pton(AF_INET,$src) || die "no IPv4 $src"),
	scalar(inet_pton(AF_INET,$dst) || die "no IPv4 $dst"),
    );

    if (defined $data) {
	if (defined $chksum_offset) {
	    my $ckdata = $no_pseudo_header ? $data :
		substr($hdr,-8).pack('xCna*',
		    $protocol,length($data),  # proto + len
		    $data
		);
	    substr($data,$chksum_offset, 2) = pack('n',ip_chksum($ckdata));
	}
	substr($hdr,10,2) = pack('n',ip_chksum($hdr));
	return $hdr.$data;
    }

    # data not defined, return sub which creates packet once data are known
    if (! defined $chksum_offset) {
	return sub {
	    substr(my $lhdr = $hdr,2,2) = pack('n',length($_[0])+20);
	    substr($lhdr,10,2) = pack('n',ip_chksum($lhdr));
	    return $lhdr.$_[0];
	};
    }
    return sub {
	my $data = shift;
	my $ckdata = $no_pseudo_header ? $data : 
	    substr($hdr,-8).pack('xCna*',
		$protocol,length($data),  # proto + len
		$data
	    );
	substr($data,$chksum_offset, 2) = pack('n',ip_chksum($ckdata));
	substr(my $lhdr = $hdr,2,2) = pack('n',length($data)+20);
	substr($lhdr,10,2) = pack('n',ip_chksum($lhdr));
	return $lhdr.$data;
    };
}

# construct IPv6 packet
sub ip6_packet {
    my ($data,$src,$dst,$protocol,$chksum_offset) = @_;
    my $hdr = pack('NnCCA16A16',
	6 << 28 | 0 << 20 | 0,       # version, traffic class, flow label
	defined($data) ? length($data) : 0,  # length of payload
	$protocol,                   # next header = protocol
	128,                         # hop limit
	scalar(inet_pton(AF_INET6,$src) || die "no IPv6 $src"),
	scalar(inet_pton(AF_INET6,$dst) || die "no IPv6 $dst"),
    );

    if (defined $data) {
	# return packet
	if (defined $chksum_offset) {
	    my $ckdata = substr($hdr,-32).pack('NxxxCa*',
		length($data), $protocol, # len + proto
		$data
	    );
	    substr($data,$chksum_offset, 2) = pack('n',ip_chksum($ckdata));
	}
	return $hdr.$data;
    }

    # data not defined, return sub which creates packet once data are known
    if (! defined $chksum_offset) {
	return sub {
	    substr($hdr,4,2) = pack('n',length($_[0]));
	    return $hdr.$_[0]
	}
    }
    return sub {
	my $data = shift;
	substr($hdr,4,2) = pack('n',length($data));
	my $ckdata = substr($hdr,-32).pack('NxxxCa*',
	    length($data), $protocol, # len + proto
	    $data
	);
	substr($data,$chksum_offset, 2) = pack('n',ip_chksum($ckdata));
	return $hdr.$data;
    };
}

sub ip_packet {
    goto &ip6_packet if $_[1] =~m{:};
    goto &ip4_packet;
}

sub ip_chksum16 {
    my $data = pop;
    $data .= "\x00" if length($data) % 2; # padding
    my $sum = 0;
    $sum += $_ for (unpack('n*', $data));
    $sum = ($sum >> 16) + ($sum & 0xffff);
    $sum = ~(($sum >> 16) + $sum) & 0xffff;
    return $sum;
}

sub ip_chksum32 {
    my $data = pop;
    $data .= "\x00" x (4 - length($data) % 4); # padding
    my $sum = 0;
    $sum += $_ for unpack('N*', $data);
    $sum = ($sum >> 16) + ($sum & 0xffff);
    $sum = ($sum >> 16) + ($sum & 0xffff);
    $sum = ($sum >> 16) + ($sum & 0xffff);
    return ~$sum;
}

require Config;
*ip_chksum = $Config::Config{ivsize} == 8 ? \&ip_chksum32 : \&ip_chksum16;

1;


lib/Net/PcapWriter/TCP.pm
use strict;
use warnings;

package Net::PcapWriter::TCP;
use fields qw(flow writer l2prefix pktmpl last_timestamp connected);

use Net::PcapWriter::IP;
use Socket qw(AF_INET IPPROTO_TCP);

sub new {
    my ($class,$writer,$src,$sport,$dst,$dport) = @_;
    my $self = fields::new($class);
    $self->{flow} = [
	# src, dst, sport, dport, state, sn
	# state = 0bFfSs: send[F]inack|send[f]in|send[S]ynack|send[s]yn
	# sn gets initialized on sending SYN
	[ $src,$dst,$sport,$dport,0,     undef ],
	[ $dst,$src,$dport,$sport,0,     undef ],
    ];
    $self->{writer} = $writer;
    $self->{last_timestamp} = undef;
    $self->{l2prefix} = $self->{writer}->layer2prefix($src);
    $self->{pktmpl} = [
	ip_packet( undef, $src, $dst, IPPROTO_TCP, 16),
	ip_packet( undef, $dst, $src, IPPROTO_TCP, 16),
    ];
    return $self;
}

sub write_with_flags {
    my ($self,$dir,$data,$flags,$timestamp) = @_;
    $flags ||= {};
    my $flow = $self->{flow}[$dir];

    if ($flags->{syn} and ($flow->[4] & 0b0001) == 0) {
	$flow->[4] |= 0b0001;
	$flow->[5] ||= rand(2**32);
    }
    if ($flags->{rst}) {
	# consider closed
	$flow->[4] |= 0b1100;
	$self->{flow}[$dir?0:1][4] |= 0b1100;
    }
    if ($flags->{ack}) {
	$flow->[4] |= 0b0010 if ($flow->[4] & 0b0011) == 0b0001; # ACK for SYN
	$flow->[4] |= 0b1000 if ($flow->[4] & 0b1100) == 0b0100; # ACK for FIN
    }

    my $sn = $flow->[5];
    if ($flags->{fin}) {
	if (($flow->[4] & 0b0100) == 0) {
	    $flow->[4] |= 0b0100;
	    $flow->[5]++
	}
    }

    return if ! defined $data; # only update state

    my $ack = $self->{flow}[$dir?0:1][5];
    $flags->{ack} = 1 if defined $ack;

    my $f = 0;
    $f |= 0b000100 if $flags->{rst};
    $f |= 0b001000 if $flags->{psh};
    $f |= 0b010000 if $flags->{ack};
    $f |= 0b100000 if $flags->{urg};
    $f |= 0b000001 if $flags->{fin};
    if ( $flags->{syn} ) {
	$f |= 0b000010;
	$sn = ($sn-1) % 2**32;
    }

    my $tcp = pack("nnNNCCnnna*",
	$flow->[2],$flow->[3],       # sport,dport
	$sn,                         # sn
	$ack||0,                     # ack
	0x50,                        # size of TCP header >> 4
	$f,                          # flags
	$flags->{window} || 2**15,   # window
	0,                           # checksum computed later
	$flags->{urg}||0,            # urg pointer
	$data                        # payload
    );

    $flow->[5] = (
	$flow->[5]
	+ length($data)
    ) % 2**32;
    $self->{last_timestamp} = $timestamp;
    $self->{writer}->packet(
	$self->{l2prefix} . $self->{pktmpl}[$dir]($tcp),
	$timestamp
    );
}

sub write {
    my ($self,$dir,$data,$timestamp) = @_;
    _connect($self,$timestamp) if ! $self->{connected};
    write_with_flags($self,$dir,$data,undef,$timestamp);
}

sub _connect {
    my ($self,$timestamp) = @_;
    my $flow = $self->{flow};
    goto done if ($flow->[1][4] & 0b11) == 0b11
	&& ($flow->[0][4] & 0b11) == 0b11;

    # client: SYN
    write_with_flags($self,0,'',{ syn => 1 },$timestamp) 
	if ($flow->[0][4] & 0b01) == 0;

    # server: SYN+ACK
    write_with_flags($self,1,'',{ 
	($flow->[1][4] & 0b01) == 0 ? ( syn => 1 ):(),
	($flow->[1][4] & 0b10) == 0 ? ( ack => 1 ):(),
    },$timestamp) if ($flow->[1][4] & 0b11) == 0;

    # client: ACK
    write_with_flags($self,0,'',{ ack => 1 },$timestamp) 
	if ($flow->[0][4] & 0b10) == 0;

    done:
    $self->{connected} = 1;
}

sub connect {
    my ($self,$timestamp) = @_;
    _connect($self,$timestamp) if ! $self->{connected};
}

sub shutdown {
    my ($self,$dir,$timestamp) = @_;
    if (($self->{flow}[$dir][4] & 0b0100) == 0) {
	_connect($self,$timestamp) if ! $self->{connected};
	write_with_flags($self,$dir,'',{ fin => 1 },$timestamp);
	write_with_flags($self,$dir ? 0:1,'',{ ack => 1 },$timestamp);
    }
}

sub close {
    my ($self,$dir,$type,$timestamp) = @_;
    my $flow = $self->{flow};

    if (!defined $type or $type eq '') {
	# simulate close only - don't write any packets
	$flow->[0][4] |= 0b1100;
	$flow->[1][4] |= 0b1100;

    } elsif ($type eq 'fin') {
	# $dir: FIN
	write_with_flags($self,$dir,'',{ fin => 1 },$timestamp)
	    if ($flow->[$dir][4] & 0b0100) == 0;

	# $odir: FIN+ACK
	my $odir = $dir?0:1;
	write_with_flags($self,$odir,'',{
	    ($flow->[$odir][4] & 0b0100) == 0 ? ( fin => 1 ):(),
	    ($flow->[$odir][4] & 0b1000) == 0 ? ( ack => 1 ):(),
	},$timestamp) if ($flow->[$odir][4] & 0b1100) == 0;

	# $dir: ACK
	write_with_flags($self,$dir,'',{ ack => 1 },$timestamp)
	    if ($flow->[$dir][4] & 0b1000) == 0;

    } elsif ($type eq 'rst') {
	# single RST and then connection is closed
	write_with_flags($self,$dir,'',{ rst => 1 },$timestamp);

    } else {
	die "only fin|rst|undef are allowed with close"
    }
}

sub ack {
    my ($self,$dir,$timestamp) = @_;
    write_with_flags($self,$dir,'',{ ack => 1 },$timestamp);
}

sub DESTROY {
    my $self = shift;
    &close($self,0,'fin',$self->{last_timestamp});
}


1;




lib/Net/PcapWriter/UDP.pm

use strict;
use warnings;
package Net::PcapWriter::UDP;
use fields qw(flow writer);
use Net::PcapWriter::IP;
use Socket qw(AF_INET IPPROTO_UDP);

sub new {
    my ($class,$writer,$src,$sport,$dst,$dport) = @_;
    my $self = fields::new($class);
    $self->{flow} = [
	# src, dst, sport, dport
	[ $src,$dst,$sport,$dport ],
	[ $dst,$src,$dport,$sport ],
    ];
    $self->{writer} = $writer;
    return $self;
}

sub write {
    my ($self,$dir,$data,$timestamp) = @_;
    my $flow = $self->{flow}[$dir];

    my $udp = pack("nnnna*",
	$flow->[2],$flow->[3],       # sport,dport
	length($data)+8,
	0,                           # checksum
	$data                        # payload
    );

    $self->{writer}->packet( ip_packet(
	$udp,
	$flow->[0],
	$flow->[1],
	IPPROTO_UDP,
	6
    ), $timestamp );
}

1;




lib/Net/PcapWriter/ICMP.pm
use strict;
use warnings;
package Net::PcapWriter::ICMP;
use fields qw(flow writer);
use Net::PcapWriter::IP;
use Socket qw(AF_INET IPPROTO_ICMP);

sub new {
    my ($class,$writer,$src,$dst,$identifier) = @_;
    my $self = fields::new($class);
    $identifier=0 if  (!defined $identifier);
    my $seq=0;
    $self->{flow} = [
        # src, dst, identifier, sequence number
        [ $src,$dst,$identifier,$seq ],
        [ $dst,$src,$identifier,$seq ],
    ];
    $self->{writer} = $writer;
    return $self;
}



sub write {
    my ($self,$dir,$code,$type,$data,$timestamp) = @_;
    my $flow = $self->{flow}[$dir];
    my $checksum = 0;
    my $identifier=$flow->[2];
    my $seq=$flow->[3];


    my $icmp = pack("CCnnna*",
        $code,$type,             # code,type
        $checksum,
        $identifier,$seq,
        $data                        # payload
    );

    $self->{writer}->packet( ip_packet(
        $icmp,
        $flow->[0],
        $flow->[1],
        IPPROTO_ICMP,
        -2,
    ), $timestamp );
}


sub ping {
   my ($self,$dir,$data,$timestamp) = @_;
   # increment seq number
   my $flow = $self->{flow}[0];
   $flow->[3]=$flow->[3]+1;
   $flow = $self->{flow}[1];
   $flow->[3]=$flow->[3]+1;

   $self->write($dir,8,0,$data,$timestamp);
}

sub pong {
   my ($self,$dir,$data,$timestamp) = @_;
   $self->write($dir,0,0,$data,$timestamp);
}



1;



lib/Net/PcapWriter/ICMP_Echo.pm
use strict;
use warnings;
package Net::PcapWriter::ICMP_Echo;
use fields qw(id src dst seq writer);
use Net::PcapWriter::IP;

sub new {
    my ($class,$writer,$src,$dst,$identifier) = @_;
    $identifier = 0 if ! defined $identifier;
    my $self = fields::new($class);
    $self->{writer} = $writer;
    $self->{id}     = $identifier;
    $self->{src}    = $src;
    $self->{dst}    = $dst;
    return $self;
}



sub _write {
    my ($self,$dir,$seq,$data,$timestamp) = @_;

    my ($src,$dst) = $dir ? @{$self}{qw(dst src)} : @{$self}{qw(src dst)};
    my $ip6 = $src =~m{:};
    my $type = $dir ?
	$ip6 ? 129 : 0 :  # echo reply
	$ip6 ? 128 : 8;   # echo request
    my $echo = pack("CCnnna*",
	$type,
	0,             # code = 0
	0,             # checksum, computed below
	$self->{id},   # identifier
	$seq,          # sequence
	$data          # payload
    );

    $self->{writer}->packet(
	# checksum at offset 2
	# for ip4 no pseudo-header will be included in checksum
	$ip6 
	    ? ip6_packet($echo,$src,$dst, 58, 2  )
	    : ip4_packet($echo,$src,$dst,  1, 2,1),
	$timestamp
    );
}


*echo_request = \&ping;
sub ping {
   my ($self,$seq,$data,$timestamp) = @_;
   $self->_write(0,$seq,$data,$timestamp);
}

*echo_response = \&pong;
sub pong {
   my ($self,$seq,$data,$timestamp) = @_;
   $self->_write(1,$seq,$data,$timestamp);
}

1;


lib/Net/PcapWriter.pm
use strict;
use warnings;
package Net::PcapWriter;
use Time::HiRes 'gettimeofday';
use Net::PcapWriter::TCP;
use Net::PcapWriter::UDP;
use Net::PcapWriter::ICMP_Echo;

our $VERSION = '0.723';

sub new {
    my ($class,$file) = @_;
    my $self = bless { fh => undef },$class;
    $self->reopen($file);
    return $self;
}

sub reopen {
    my ($self,$file) = @_;
    my $fh;
    if ( $file ) {
	if ( ref($file)) {
	    $fh = $file
	} else {
	    open($fh,'>',$file) or die "open $file: $!";
	    binmode($fh);
	}
    } else {
	$fh = \*STDOUT;
    }
    $self->{fh} = $fh;
    $self->_header;
}

# write pcap header
sub _header {
    my $self = shift;

    # struct pcap_file_header {
    #     bpf_u_int32 magic;
    #     u_short version_major;
    #     u_short version_minor;
    #     bpf_int32 thiszone; /* gmt to local correction */
    #     bpf_u_int32 sigfigs;    /* accuracy of timestamps */
    #     bpf_u_int32 snaplen;    /* max length saved portion of each pkt */
    #     bpf_u_int32 linktype;   /* data link type (LINKTYPE_*) */
    # };

    print {$self->{fh}} pack('LSSlLLL',
	0xa1b2c3d4, # magic
	2,4,        # major, minor
	0,0,        # timestamps correction and accuracy
	0xffff,     # snaplen
	1,          # DLT_EN10MB
    );
}

sub layer2prefix {
    my $ip = pop;
    return pack("NnNnn",
	0,1,0,1, # all macs 0:*
	$ip =~m{:} ? 0x86dd: 0x0800, # ETH_TYPE_IP6 | ETH_TYPE_IP
    );
}

# write pcap packet
sub packet {
    my ($self,$data,$timestamp) = @_;
    $timestamp ||= [ gettimeofday() ];

    # struct pcap_pkthdr {
    #     struct timeval ts;  /* time stamp */
    #     bpf_u_int32 caplen; /* length of portion present */
    #     bpf_u_int32 len;    /* length this packet (off wire) */
    # };

    my ($tsec,$tmsec);
    if (ref($timestamp)) {
	# array like in Time::HiRes
	($tsec,$tmsec) = @$timestamp; 
    } else {
	$tsec = int($timestamp);
	$tmsec = int(($timestamp - $tsec)*1_000_000);
    }

    print {$self->{fh}} pack('LLLLa*',
	$tsec,$tmsec,       # struct timeval ts
	length($data),      # caplen
	length($data),      # len
	$data,              # data
    );
}


# return new TCP connection object
sub tcp_conn {
    my ($self,$src,$sport,$dst,$dport) = @_;
    return Net::PcapWriter::TCP->new($self,$src,$sport,$dst,$dport);
}

# return new UDP connection object
sub udp_conn {
    my ($self,$src,$sport,$dst,$dport) = @_;
    return Net::PcapWriter::UDP->new($self,$src,$sport,$dst,$dport);
}

# return new ICMP_Echo "connection" object
sub icmp_echo_conn {
    my ($self,$src,$dst,$id) = @_;
    return Net::PcapWriter::ICMP_Echo->new($self,$src,$dst,$id);
}

1;

__END__

=head1 NAME

Net::PcapWriter - simple creation of pcap files from code

=head1 SYNOPSIS

 use Net::PcapWriter;

 # disabling checksum calculation leads to huge performance boost
 Net::PcapWriter::IP->calculate_checksums(0);

 my $writer = Net::PcapWriter->new('test.pcap');
 my $conn = $writer->tcp_conn('1.2.3.4',1234,'5.6.7.8',80);

 # this will automatically add syn..synack..ack handshake to pcap
 # each write will be a single packet
 $conn->write(0,"POST / HTTP/1.0\r\nContent-length: 3\r\n\r\n");
 $conn->ack(1); # force ack from server

 # send another packet w/o forcing ack
 $conn->write(0,"abc");

 # client will no longer write
 $conn->shutdown(0);

 # this will automatically add ack to last packet
 $conn->write(1,"HTTP/1.0 200 Ok\r\nContent-length: 10\r\n\r\n");
 $conn->write(1,"0123456789");

 # will automatically add remaining FIN+ACK
 undef $conn;

 # write some UDP packets with IPv6
 $conn = $writer->udp_conn('dead::beaf',1234,'beaf::dead',53);
 $conn->write(0,"....");
 $conn->write(1,"....");

 # write a ping exchange (works also with IPv6)
 $conn = $writer->icmp_echo_conn('1.2.3.4','5.6.7.8',10);
 $conn->ping(1,"foo");
 $conn->ping(2,"bar");
 $conn->pong(1,"foo");


=head1 DESCRIPTION

With L<Net::PcapWriter> it is possible to create pcap files within a program
without capturing any data. This is useful for setting up test data without
setting up the needed infrastructure for data creation and capturing.

The following methods are supported:

=over 4

=item $class->new([$filename|$handle])

Creates new object.
If file name is given it will be opened for writing, if file handle is given it
will be used. Otherwise the pcap data will be written to STDOUT.
Will write pcap header for DLT_RAW to pcap file.

=item $class->reopen([$filename|$handle])

This will close the current pcap file and open a new one. No connections will
be implicitely closed, i.e. they will continue inside the new pcap file. The
main purpose is to be able to rotate the file after some time or size.

=item $writer->packet($pkt,[$timestamp])

Will write raw Layer 2 packet $pkt with $timestamp in pcap file.
$timestamp can be C<time_t> (seconds), float (like C<time_t>, but with higher
resolution) or C<<[$sec,$msec]>> like in C<<struct timeval>>.
If $timestamp is not given will use C<Time::HiRes::gettimeofday>.

To get the Layer 2 prefix in case of IP data use C<$writer->layer2prefix($ip)>.

=item $writer->tcp_conn($src,$sport,$dst,$dport)

Will return C<Net::PcapWriter::TCP> object, which then provides the following
methods:

=over 8

=item $tcpconn->write($dir,$data,[$timestamp])

Will write the given data for the direction C<$dir> (0 are data from client to
server, 1 the other way). Will write TCP handshake if not done yet.

=item $tcpconn->ack($dir,[$timestamp])

Will write an empty message with an ACK from direction C<$dir>.

=item $tcpconn->shutdown($dir,[$timestamp])

Will add FIN+ACK for shutdown from direction C<$dir> unless already done.

=item $tcpconn->write_with_flags($dir,$data,\%flags,[$timestamp])

Write a TCP packet with specific flags, like 
C<<{ syn => 1, ack => 1 }>>. This is also internally used to
automatically add the initial handshake (i.e SYN from client, SYN+ACK from
server and SYN+ACK from client) and the close of the connection (FIN), whereby
the close can be easier handled with C<shutdown>.

Possible flags are syn, ack, fin, rst, psh and rst.

If C<$data> is undef only the internal state regarding the flags will be set.

=item $tcpconn->close($dir,$flag,[$timestamp]);

Close the connection. C<$dir> is the side which initiates the close and C<$flag>
is how the connection is closed, i.e. C<'fin'>, C<'rst'> or C<''>. In the last
case no data will be written but only the internal state will be set to mark the
connection as closed. This way no more closing data will be written on DESTROY
of the object.

Note that the DESTROY of the object will automatically write a normal close
(with FIN) if the connection is not yet considered closed and thus an explicit
close is only needed if one needs more explicit control how the closing should
look like.

=item $tcpconn->connect($dir,[$timestamp])

This will explicitely open the connection (3-way handshake) unless it is already
open. This will be implicitely done if the connection is not fully open in case
of C<write> or C<shutdown> so usually it is not necessary to call this function
explicitely.

=item undef $tcpconn

Will call shutdown for both C<$dir> before destroying connection object.

=back

=item $writer->udp_conn($src,$sport,$dst,$dport)

Will return C<Net::PcapWriter::UDP> object, which then provides the following
methods:

=item $tcpconn->write($dir,$data,[$timestamp])

Will write the given data for the direction C<$dir> (0 are data from client to
server, 1 the other way).

=item $writer->icmp_echo_conn($src,$dst,[$id])

Will return C<Net::PcapWriter::ICMP_Echo> object which provides a connection
with echo request and reply using the identifier $id (default 0). This object
can handle echo request/reply for ICMP and ICMPv6.
It has the following methods:

=item $echo->ping($seq,$data,[$timestamp])

Will write an ICMP echo request from connection source to destination with
sequence $seq and data $data.

=item $echo->pong($seq,$data,[$timestamp])

Will write an ICMP echo reply from connection destination to source with
sequence $seq and data $data.

=back

Additionally a huge performance boost can be reached by disabling checksum
calculation:

   Net::PcapWriter::IP->calculate_checksums(0);

=head1 AUTHOR

Steffen Ullrich <sullr@cpan.org>

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2015 by Steffen Ullrich.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.
